### PR TITLE
python312Packages.greatfet: add missing dependencies

### PR DESCRIPTION
**Pull request by msanft**
## Description of changes

Adds missing dependencies and some fixups to the greatfet package.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/greatfet/default.nix [hunk 1]
```diff
@@ -20,15 +25,22 @@ buildPythonPackage rec {
     sha256 = "sha256-AKpaJZJTzMY3IQXLvVnLWh3IHeGp759z6tvaBl28BHQ=";
```
Comment (dotlambda) [inline pkgs/development/python-modules/greatfet/default.nix:25]:
```suggestion
    hash = "sha256-AKpaJZJTzMY3IQXLvVnLWh3IHeGp759z6tvaBl28BHQ=";
```

**Review by dotlambda — COMMENTED** [review]
Please use `pyproject = true` and set `build-system`.

File: pkgs/development/python-modules/greatfet/default.nix [hunk 2]
```diff
@@ -20,15 +25,22 @@ buildPythonPackage rec {
...
   };
 
-  disabled = !isPy3k;
+  disabled = pythonOlder "3.8";
 
   propagatedBuildInputs = [
```
Comment (dotlambda) [inline pkgs/development/python-modules/greatfet/default.nix:30]:
```suggestion
  dependencies = [
```

File: pkgs/development/python-modules/greatfet/default.nix [hunk 3]
```diff
@@ -1,34 +1,54 @@
...
+    pyusb
+    tabulate
+    tqdm
   ];
 
+  # Tests seem to require devices (or simulators) which are
+  # not available in the build sandbox.
   doCheck = false;
 
   preBuild = ''
```
Comment (dotlambda) [inline pkgs/development/python-modules/greatfet/default.nix:54]:
```suggestion
  postPatch = ''
```

File: pkgs/development/python-modules/greatfet/default.nix [hunk 4]
```diff
@@ -1,34 +1,54 @@
...
-    sha256 = "sha256-AKpaJZJTzMY3IQXLvVnLWh3IHeGp759z6tvaBl28BHQ=";
+    hash = "sha256-AKpaJZJTzMY3IQXLvVnLWh3IHeGp759z6tvaBl28BHQ=";
   };
 
-  disabled = !isPy3k;
+  disabled = pythonOlder "3.8";
+
+  build-system = [
+    setuptools
+    setuptools-git-versioning
```
Comment (dotlambda) [inline pkgs/development/python-modules/greatfet/default.nix:35]:
```suggestion
```
Let's patch that out because we set the version manually.

Comment (msanft) [reply pkgs/development/python-modules/greatfet/default.nix:35]: It doesn't complain if it's not there. So I guess we can just leave it as is

File: pkgs/development/python-modules/greatfet/default.nix (hunk 3 above)
Comment (dotlambda) [reply pkgs/development/python-modules/greatfet/default.nix:50]: You could also use `sourceRoot`.

Comment (msanft) [reply pkgs/development/python-modules/greatfet/default.nix:50]: I don't think the Python builder supports that?

Comment (msanft) [reply pkgs/development/python-modules/greatfet/default.nix:50]: That disables access to `..`, which we need to write the version

**Comment by msanft** [comment]
It does indeed now complain because of setuptools-git-versioning. Preparing a patch

**Comment by msanft** [comment]
Result of `nixpkgs-review pr 331055` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.greatfet</li>
    <li>python311Packages.greatfet.dist</li>
    <li>python312Packages.greatfet</li>
    <li>python312Packages.greatfet.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/greatfet/default.nix (hunk 3 above)
Comment (dotlambda) [reply pkgs/development/python-modules/greatfet/default.nix]: No it doesn't.

Comment (msanft) [reply pkgs/development/python-modules/greatfet/default.nix]: Shouldn't matter now, as of the test

But when trying to access `..` with `sourceRoot` set, it just pointed into `/nix/store`

Comment (dotlambda) [reply pkgs/development/python-modules/greatfet/default.nix]: It does point to the correct directory but only with read access.

Comment (msanft) [reply pkgs/development/python-modules/greatfet/default.nix]: Ok, maybe I made a mistake somewhere, not sure.
But as I said, it shouldn't matter now

**Comment by dotlambda** [comment]
I replaced the patch with subtituteInPlace because there's no need to add 80 lines to nixpkgs.

**Review by dotlambda — APPROVED** [review]

**Comment by msanft** [comment]
@dotlambda would you mind discussing to remove the patch the next time? I don't find it very nice to just "destroy" work done by someone else without any discussion, even it it does the same in the end

**Comment by dotlambda** [comment]
> @dotlambda would you mind discussing to remove the patch the next time? I don't find it very nice to just "destroy" work done by someone else without any discussion, even it it does the same in the end

It would have taken longer to keep reviewing. I just copy-pasted this from other packages that patch out setuptools-git-versioning.

**Pull request merged**